In [1]:
# imports
from tqdm import tqdm
import pandas as pd
import numpy as np
import os

# Gather the Data
First step is to gather the data. There is a module defined called RetroSheetDownloader with a class called DownloadRetroSheetData. Using the class, call the run method that will download the data in .txt format.

In [2]:
from RetroSheetDownloader import DownloadRetroSheetData

# Create an instance of the class and calls the run method. By default, we download the data from the years 2014 to 2019, inclusive 
DownloadRetroSheetData().run()

Unzipping gl2019.zip: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:06<00:00,  1.03s/log]


Next, let's try opening the data with pandas so we can see it in tabular form.

In [3]:
import pandas as pd
import os

RETROSHEET_FILES = "./datasets/retro_sheet_gls"

# We will be looking at the 2014 data as an example
retro_sheet_data_gl2014 = pd.read_csv("./datasets/retro_sheet_gls/GL2014.TXT")

# Get all the games that the Yankees played in 2014
nya_gl_2014 = retro_sheet_data_gl2014.loc[(retro_sheet_data_gl2014["Home Team"] == 'NYA') | 
                                          (retro_sheet_data_gl2014["Visiting Team"] == 'NYA')]

# Sort the games played by date
display(nya_gl_2014.sort_values("Date"))

Date  Number of game  Day Visiting Team Visiting Team League  \
17    20140401               0  Tue           NYA                   AL   
28    20140402               0  Wed           NYA                   AL   
42    20140403               0  Thu           NYA                   AL   
56    20140404               0  Fri           NYA                   AL   
71    20140405               0  Sat           NYA                   AL   
...        ...             ...  ...           ...                  ...   
2362  20140924               0  Wed           BAL                   AL   
2375  20140925               0  Thu           BAL                   AL   
2384  20140926               0  Fri           NYA                   AL   
2400  20140927               0  Sat           NYA                   AL   
2415  20140928               0  Sun           NYA                   AL   

      Visiting Team Game Number Home Team Home Team League  \
17                            1       HOU               AL   
28                            2       HOU               AL   
42                            3       HOU               AL   
56                            4       TOR               AL   
71                            5       TOR               AL   
...                         ...       ...              ...   
2362                        158       NYA               AL   
2375                        159       NYA               AL   
2384                        160       BOS               AL   
2400                        161       BOS               AL   
2415                        162       BOS               AL   

      Home Team Game Number  Visiting Team Score  ...  \
17                        1                    2  ...   
28                        2                    1  ...   
42                        3                    4  ...   
56                        5                    7  ...   
71                        6                    0  ...   
...                     ...                  ...  ...   
2362                    158                    9  ...   
2375                    159                    5  ...   
2384                    160                    3  ...   
2400                    161                    4  ...   
2415                    162                    9  ...   

      Home Team Player 7 Name  Home Team Player 7 Defensive Position  \
17             Matt Dominguez                                      5   
28             Matt Dominguez                                      5   
42             Matt Dominguez                                      5   
56               Brett Lawrie                                      5   
71                 Josh Thole                                      2   
...                       ...                                    ...   
2362              Chris Young                                      7   
2375             Stephen Drew                                      4   
2384             Bryce Brentz                                      7   
2400          Rusney Castillo                                      8   
2415             Bryce Brentz                                      7   

     Home Team Player 8 ID Home Team Player 8 Name  \
17                hoesl001               L.J. Hoes   
28                presa001            Alex Presley   
42                presa001            Alex Presley   
56                goinr001              Ryan Goins   
71                goinr001              Ryan Goins   
...                    ...                     ...   
2362              richa001       Antoan Richardson   
2375              suzui001           Ichiro Suzuki   
2384              lavar001          Ryan Lavarnway   
2400              brenb002            Bryce Brentz   
2415              rossd001              David Ross   

      Home Team Player 8 Defensive Position Home Team Player 9 ID  \
17                                        9              villj001   
28                                     

Using this idea we can generate the starting lineup for each team for each season's game logs that we download. Let's first get all of the teams in the league. We can keep a set of the team abbreviations. We can then loop through all of the game log files we got from *retrosheet* and for each team, make a dataframe that has all of games for one particular team.

In [4]:
# We will store the games played by each team here
TEAM_GAMES = "./datasets/games_by_teams/"

# There should be 30 teams in the MLB
teams = set(retro_sheet_data_gl2014["Home Team"])
assert len(teams) == 30

def generate_all_games_per_team(all_game_per_season, all_games_per_team_dir, teams):
    """For each team, generate the games that they have played over the course of the years"""
    
    if not os.path.exists(all_games_per_team_dir):
        os.mkdir(all_games_per_team_dir)
    
    # Delete the files that are in the folder already just in case
    for f in os.listdir(all_games_per_team_dir):
        os.remove(os.path.join(all_games_per_team_dir, f))

    with tqdm(total=len(teams), unit="teams") as pbar:
        for team in teams:
            pbar.set_description(team)
            team_dfs = list()
            for file in os.listdir(all_game_per_season):
                file_name = os.path.join(all_game_per_season, file)
                gl_year = pd.read_csv(file_name)

                # Get the tables for the team's home and away games
                team_gl_year = gl_year.loc[(gl_year["Home Team"] == team) |
                                            (gl_year["Visiting Team"] == team)]

                # Add the concatenated dataframe to the placeholder list
                team_dfs.append(team_gl_year)

            # Concatenate all of the intermediate 
            team_all_games = pd.concat(team_dfs).sort_values("Date")
            team_all_games.to_csv(os.path.join(all_games_per_team_dir, f"{team}_all_games.csv"), index=None)
            pbar.update()

generate_all_games_per_team(RETROSHEET_FILES, TEAM_GAMES,teams)

SEA: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 30/30 [00:21<00:00,  1.38teams/s]


Let's quickly look at an arbitrary csv file we just generated and make sure that there are 30 new files generated.

In [5]:
# Load Tampa Bay Rays as an example
tba_gl_all = pd.read_csv(os.path.join(TEAM_GAMES, "TBA_all_games.csv"))

# There are 162 games per season and the number of seasons that we have are the number of
# retrosheet files that we downloaded
assert tba_gl_all.shape[0] == 162 * len([f for f in os.listdir(RETROSHEET_FILES)])

display(tba_gl_all)

Date  Number of game  Day Visiting Team Visiting Team League  \
0    20140331               0  Mon           TOR                   AL   
1    20140401               0  Tue           TOR                   AL   
2    20140402               0  Wed           TOR                   AL   
3    20140403               0  Thu           TOR                   AL   
4    20140404               0  Fri           TEX                   AL   
..        ...             ...  ...           ...                  ...   
967  20190924               0  Tue           NYA                   AL   
968  20190925               0  Wed           NYA                   AL   
969  20190927               0  Fri           TBA                   AL   
970  20190928               0  Sat           TBA                   AL   
971  20190929               0  Sun           TBA                   AL   

     Visiting Team Game Number Home Team Home Team League  \
0                            1       TBA               AL   
1                            2       TBA               AL   
2                            3       TBA               AL   
3                            4       TBA               AL   
4                            4       TBA               AL   
..                         ...       ...              ...   
967                        158       TBA               AL   
968                        159       TBA               AL   
969                        160       TOR               AL   
970                        161       TOR               AL   
971                        162       TOR               AL   

     Home Team Game Number  Visiting Team Score  ...  Home Team Player 7 Name  \
0                        1                    2  ...               Matt Joyce   
1                        2                    4  ...           Logan Forsythe   
2                        3                    3  ...           Sean Rodriguez   
3                        4                    2  ...               Matt Joyce   
4                        5                    1  ...            Brandon Guyer   
..                     ...                  ...  ...                      ...   
967                    158                    1  ...              Joey Wendle   
968                    159                    0  ...               Matt Duffy   
969                    160                    6  ...        Teoscar Hernandez   
970                    161                    1  ...           Anthony Alford   
971                    162                    3  ...            Richard Urena   

     Home Team Player 7 Defensive Position Home Team Player 8 ID  \
0                                       10              molij001   
1                                       10              hanir001   
2                                        7              molij001   
3                                        9              hanir001   
4                                        7              hanir001   
..                                     ...                   ...   
967                                      4              adamw002   
968                                      5              kierk001   
969                                      8              urenr001   
970                                      7              jansd001   
971                                      6              davij007   

     Home Team Player 8 Name  Home Team Player 8 Defensive Position  \
0                Jose Molina                                      2   
1               Ryan Hanigan                                      2   
2                Jose Molina                                      2   
3               Ryan Hanigan                                      2   
4               Ryan Hanigan                                      2   
..                       ...                                    ...   
967             Willy Adames                                      6   
968          Kevin Kiermaier             

# Preliminary Observations

## Does Being the Home Team Matter?

Alright, so at this point we might be able to look at couple of things that might indicate as being good labels to use in our models. One thing we might look at is how often home teams win their games. It's thought to be true that there is an inherent advantage of teams playing in front of their own fans. Let's see if there is any correlation. If being the home team matters, then the number of home games won over the number of total games won should be over 50%.

In [6]:
def win_percentage_of_home_games(games_dir):
    """Function which returns the percentage of games won by the home team"""
    
    df_list = []
    for file in os.listdir(games_dir):
        df_list.append(pd.read_csv(os.path.join(games_dir, file)))
    all_games = pd.concat(df_list)
    home_game_wins = all_games.loc[all_games["Home Team Score"] > all_games["Visiting Team Score"]].shape[0]
    return home_game_wins/all_games.shape[0]

win_percentage_of_home_games(RETROSHEET_FILES)

0.5331000891815875

As we can see, out of all the games played, the home team only won 53% of the time, meaning that there doesn't seem to be an inherit advantage for the home team. This is honestly lower than we expected and are quite surprised with the result. There are a few things to note here however. This percentage is based on the whole league. It might be smarter to look at two more percentages:
1. The number of games won at home out of the total home games **per team**
2. The number of games won at home out of the total games won **per team**

## Number of Runs Scored vs. Wins

Naturally a more potent offense should win more games. And generally, a better offense will score more runs. We should see whether or not a team that scores more runs will win more games